ice-adapter/src/main/java/com/faforever/iceadapter/util/latencyestimation/FafLatencyEstimator.java

```diff
@@ -0,0 +1,92 @@
+package com.faforever.iceadapter.util.latencyestimation;
+
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
+
+@Slf4j
+public class FafLatencyEstimator {
+
+    private static final int FAF_LATENCY_ESTIMATION_RETRIES = 3;
+    private static final int FAF_LATENCY_ESTIMATION_RETRY_PERIOD = 500;
+    public static final String FAF_REQUEST = "{\"command\":\"ping\"}\n";
+
+
+    // estimate latency multiple times and take lowest to compensate for server processing time
+    public static CompletableFuture<Double> getLatency(String hostname) {
+        CompletableFuture<Double> future = new CompletableFuture<>();
+
+        log.info("Estimating latency to {} using fafp", hostname);
+
+        new Thread(() -> {
+            List<CompletableFuture<Double>> estimations = new ArrayList<>();
+            for(int i = 0;i < FAF_LATENCY_ESTIMATION_RETRIES;i++) {
+                estimations.add(pingFafServer(hostname));
+                try { Thread.sleep(FAF_LATENCY_ESTIMATION_RETRY_PERIOD); } catch(InterruptedException ignored) {}
+            }
+            CompletableFuture.allOf(estimations.toArray(new CompletableFuture[0])).thenRun(() -> {
+                double lowest = estimations.stream().mapToDouble(doubleCompletableFuture -> {
+                    try {
+                        return doubleCompletableFuture.get();
+                    } catch (ExecutionException | InterruptedException e) {
+                        log.error("Error while obtaining latency for faf server", e);
+                        return 10000;
+                    }
+                }).min().orElse(10000);
+                future.complete(lowest);
+            });
+            try { Thread.sleep(3000); } catch(InterruptedException e) {}
+            if(! future.isDone()) {
+                future.cancel(false);
+            }
+        }).start();
+
+        return future;
```

Comment:
Why launch a thread like this instead of using something like CompletableFuture.anyOf() or future.applyToEither where you supply an timeout future with a predetermined duration?